File: driver/go/connection.go
/*
 * Copyright (c) 2019 Google LLC.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package sfdb

import (
	"context"
	"database/sql/driver"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"

	sfdb_pb "github.com/googlegsa/sfdb/grpc_sfdb_service_go_proto"
	"google.golang.org/grpc"
	"google.golang.org/grpc/connectivity"
)

var (
	defaultTTLSeconds = 5
)

// Connection struct of SFDB driver RPC conn.
type Connection struct {
	rpcConn *grpc.ClientConn
	// username:password
	creds string
	// ipv4:port, ipv6:port, hostname:port
	addr   string
	dbName string
	// parsed to DBParams struct: param1=value1&...&paramN=valueN
	params *DBParams
	// RPC stub object pointer
	stub *sfdb_pb.SfdbServiceClient
}

// DBParams contains structured fields of db params
type DBParams struct {
	// TTL for RPC request.
	// This should be <= `deadline` protobuf field if it's set.
	ttl int
	// If true, then SFDB returns protobuf response as
	// stringified JSON format (on SELECT queries),
	// if false, then SFDB returns protobuf with the serialized, compressed
	// protobufs, and SFDB driver will de-serialize into Row structs.
	includeDebugStrings bool
}

// Parse dbParams part of connection string
func (p *DBParams) Parse(dbParams string) error {
	params := strings.Split(dbParams, "&")

	for i := 0; i < len(params); i++ {
		pair := strings.Split(params[i], "=")
		if len(pair) != 2 {
			continue
		}
		k, v := pair[0], pair[1]

		switch k {
		case "ttl":
			ttl, err := strconv.ParseInt(v, 10, 32)
			if err != nil {
				return err
			}
			// convert int64 to int
			p.ttl = int(ttl)
		case "includeDebugStrings":
			flag, err := strconv.ParseBool(v)
			if err != nil {
				return err
			}
			p.includeDebugStrings = flag
		}
	}

	return nil
}

// Begin function starts a transaction.
func (c *Connection) Begin() (driver.Tx, error) {
	err := errors.New("sfdb does not support transactions")
	return nil, err
}

// Prepare returns a prepared statement, bound to this Connection.
func (c *Connection) Prepare(query string) (driver.Stmt, error) {
	return c.PrepareContext(context.Background(), query)
}

func (c *Connection) PrepareContext(ctx context.Context, query string) (driver.Stmt, error) {
	if err := sanitizeQuery(query); err != nil {
		return nil, err
	}
	s := &stmt{
		conn:  c,
		query: query,
	}
	return s, nil
}

// Close RPC connection with SFDB target.
func (c *Connection) Close() error {
	if c.rpcConn != nil {
		fmt.Println("Closing RPC connection..")
		c.rpcConn.Close()
		c.rpcConn = nil
	}
	return nil
}

func (c *Connection) QueryContext(ctx context.Context, query string,
	args []driver.NamedValue) (driver.Rows, error) {
	processedQuery, err := interpolate(query, args)
	if err != nil {
		return nil, err
	}
	return QueryRPC(ctx, processedQuery, c)
}

func (c *Connection) ExecContext(ctx context.Context, query string,
	args []driver.NamedValue) (driver.Result, error) {
	processedQuery, err := interpolate(query, args)
	if err != nil {
		return nil, err
	}
	return ExecRPC(ctx, processedQuery, c)
}

// ParseConnString parses connection string to Connection struct.
// Example of Connection string:
// [username[:password]@][address]/dbname[?param1=value1&...&paramN=valueN]
func ParseConnString(c *Connection, connString string) error {
	tmp := strings.TrimSpace(connString)

	if len(tmp) == 0 {
		err := errors.New("empty Connection string")
		return err
	}

	// username:password
	creds := strings.Split(tmp, "@")
	if len(creds) > 2 {
		err := errors.New("too many @ symbols")
		return err
	} else if len(creds) == 2 {
		c.creds = creds[0]
		tmp = creds[1]
	}

	// addr:port
	addrs := strings.Split(tmp, "/")
	if len(addrs) > 2 {
		err := errors.New("too many / symbols")
		return err
	} else {
		c.addr = addrs[0]
		if c.addr == "" {
			err := errors.New("invalid address")
			return err
		}
		if len(addrs) == 2 {
			tmp = addrs[1]
		} else {
			tmp = ""
		}
	}

	// dbname
	dbname := strings.Split(tmp, "?")
	if len(dbname) > 2 {
		err := errors.New("too many ? symbols")
		return err
	} else if len(dbname) == 2 {
		c.dbName = dbname[0]
		// db params
		err := c.params.Parse(dbname[1])
		if err != nil {
			return err
		}
	} else {
		c.dbName = tmp
	}

	return nil
}

func newConnection() *Connection {
	return &Connection{
		params: &DBParams{ttl: defaultTTLSeconds},
	}
}

// Establishes actual connection to GRPC server
func ConnectToService(address string, timeout time.Duration) (*grpc.ClientConn, *sfdb_pb.SfdbServiceClient, error) {
	rpcConn, err := grpc.Dial(address, grpc.WithBlock(), grpc.WithTimeout(timeout), grpc.WithInsecure())
	if err != nil {
		return nil, nil, err
	}

	stub := sfdb_pb.NewSfdbServiceClient(rpcConn)
	return rpcConn, &stub, nil
}

// Connect establishes RPC Connection with SFDB target host
// with the given Connection string.
// TODO: Use a pool of connections?
func Connect(connString string) (*Connection, error) {
	var err error
	conn := newConnection()

	// 1. parse Connection string
	err = ParseConnString(conn, connString)
	if err != nil {
		return nil, err
	}

	// 2. establish RPC Connection
	conn.rpcConn, conn.stub, err = ConnectToService(conn.addr, 1*time.Second)

	return conn, err
}

// Ping implements driver.Pinger interface
func (c *Connection) Ping(ctx context.Context) (err error) {
	if c.rpcConn == nil {
		return driver.ErrBadConn
	}

	if c.rpcConn.GetState() != connectivity.Ready && c.rpcConn.GetState() != connectivity.Idle {
		return driver.ErrBadConn
	}

	return nil
}

func (c *Connection) Redirect(new_host string) (err error) {
	parts := strings.Split(new_host, ":")
	if len(parts) < 2 {
		return errors.New(fmt.Sprintf("Invalid redirect address: %s", new_host))
	}

	var new_service_address = parts[0] + ":" + parts[1]

	rpcConn, stub, err := ConnectToService(new_service_address, 1*time.Second)

	if err != nil {
		return err
	}

	c.rpcConn = rpcConn
	c.stub = stub

	return nil
}

File: driver/python/connection.py
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

"""Connection for the Google SFDB DB-API."""

import grpc
import logging
import sys
import platform
import exceptions
from cursor import Cursor
from logging import getLogger

from server import grpc_service_pb2_grpc
from server import grpc_service_pb2

TIMEOUT = 5
PYTHON_VERSION = ".".join([str(v) for v in sys.version_info[:3]])
PLATFORM = platform.platform()
_LOGGER = getLogger(__name__)


class Connection(object):
    """DB-API connection to SFDB"""

    def __init__(self, host_port, dbname, params, creds):
        """Creates and maintains gRPC channel.

        Args:
            host_port(tuple(str, str)): hostname:port 
            dbname(str): database name to connect (not implemented in SFDB)
            params(*str): connection parameters
            creds(tuple(str, str)): username:password

        Returns:
            Connection object

        Raises:
            InterfaceError: if there's any gRPC channel related exception
        """
        self.addr, self.port = host_port  # (ipv4:port|ipv6:port|hostname:port)
        self._creds = creds
        self._dbname = dbname
        self._params = params
        self._timeout = TIMEOUT
        self._grpc_channel = None
        self._stub = None
        self._errors = {}

        _LOGGER.debug(u"SFDB connector. Python Version: %s, Platform: %s",
                      PYTHON_VERSION, PLATFORM)
        self._connect()

    def _connect(self):
        try:
            self._grpc_channel = grpc.insecure_channel(
                "{}:{}".format(self.addr, self.port))
            self._stub = grpc_service_pb2_grpc.SfdbServiceStub(
                self._grpc_channel)
            _LOGGER.debug("gRPC channel and stub initialized")
        except grpc.RpcError as rpc_error:
            err_msg = "{}: {}".format(
                rpc_error.code().name, rpc_error.details())
            _LOGGER.error("gRPC channel error %s", err_msg)
            raise exceptions.InterfaceError(err_msg)

    def close(self):
        """Close the connection now.

        Connection will not be usable and will raise exception."""
        self._stub = None
        self._grpc_channel.close()
        _LOGGER.info("gRPC connection closed.")

    def commit(self):
        """SFDB does not support transactions."""
        pass

    def rollback(self):
        """SFDB does not support transactions."""
        pass

    def cursor(self, cursor_class=Cursor):
        """Creates cursor object using the connection.

        Each statement should create a new cursor object."""
        _LOGGER.debug("Creating cursor.")
        if not self._grpc_channel:
            raise exceptions.InterfaceError()
        return cursor_class(self)

    def cmd_query(self, req):
        """Executes SQL protobuf over gRPC.

        Args:
            req: request protobuf

        Returns:
            responce protubuf

        Raises: 
            InterfaceError if there are any errors.
        """
        try:
            resp = self._stub.ExecSql(req)
        except grpc.RpcError as rpc_error:
            resp = str(rpc_error)
            self._errors = {"status_code": rpc_error.code().name,
                            "details": rpc_error.details()}
            _LOGGER.error("Error during request: %s", self._errors)
            raise exceptions.InterfaceError(self._errors)
        return resp

    def _redirect(self, new_sock):
        try:
            self.addr, self.port, _ = new_sock.split(":")
        except ValueError as e:
            raise exceptions.InterfaceError("Invalid redirect address:", e)
        self._connect()

    def __repr__(self):
        return str(list(map(str, (self.addr,
                                  self.port,
                                  self._dbname,
                                  self._params))))


def connect(conn_str):
    """Construct a DB-API connection to SFDB.

    Args:
        conn_str(str): string containing connection attributes

    Returns:
        Connection object

    Raises:
        None
    """
    addr, dbname, params, creds = _parse_conn_string(conn_str)
    return Connection(addr, dbname=dbname, params=params, creds=creds)


def _parse_conn_string(conn_str):
    """Parses connection string to Connection.

    Example of connection string:
    [username[:password]@]address/dbname[?param1=value1&...&paramN=valueN]
    returns: creds dict, address tuple, dbname, params dict

    Args:
        conn_str(str): string containing connection attributes

    Returns:
        tuple containing parsed connection attributes
    """
    conn_str = conn_str.strip()
    if not conn_str:
        raise exceptions.InterfaceError("Empty connection string.")
    if r'@' in conn_str:
        creds_str, conn_str = conn_str.split(r'@')
        if r':' in creds_str:
            username, password = creds_str.split(r':')
        else:
            username, password = creds_str, None
        creds = {username: password}
    else:
        creds = None
    if r'?' in conn_str:
        conn_str, params_str = conn_str.split(r'?')
        params = {}
        for param in params_str.split(r'&'):
            k, v = param.split(r'=')
            params[k] = v
    else:
        params = None
    conn_str, dbname = conn_str.split(r'/')
    sock = tuple(conn_str.split(r':'))
    return sock, dbname, params, creds
